src/dune_rules/coq_rules.ml

```diff
@@ -581,103 +407,195 @@ let source_rule ~sctx theories =
+  let* scope = Scope.DB.find_by_dir dir in
+  let theory =
+    let coq_lib_db = Scope.coq_libs scope in
+    Coq_lib.DB.resolve coq_lib_db ~coq_lang_version:s.buildable.coq_lang_version
+      s.name
+  in
   let theories_deps =
     Resolve.Memo.bind theory ~f:(fun theory ->
         Resolve.Memo.lift @@ Coq_lib.theories_closure theory)
   in
+  (* ML-level flags for depending libraries *)
+  let ml_flags, mlpack_rule =
+    let lib_db = Scope.libs scope in
+    plugins_of_buildable ~context ~theories_deps ~lib_db s.buildable
+  in
+  let wrapper_name = Coq_lib_name.wrapper (snd s.name) in
+  let name = snd s.name in
+  let loc = s.buildable.loc in
+  let use_stdlib = s.buildable.use_stdlib in
+  let stanza_flags = s.buildable.flags in
   let* coq_dir_contents = Dir_contents.coq dir_contents in
   let theory_dirs =
     Coq_sources.directories coq_dir_contents ~name |> Path.Build.Set.of_list
   in
-  let coqc_dir = (Super_context.context sctx).build_dir in
-  let+ cctx =
-    Context.create sctx ~coqc_dir ~dir ~wrapper_name ~theories_deps ~theory_dirs
-      s.buildable
-  in
   let coq_modules = Coq_sources.library coq_dir_contents ~name in
-  (cctx, coq_modules)
-
-let setup_vo_rules ~sctx ~dir ~(cctx : _ Context.t) (s : Theory.t) theory
-    coq_modules =
-  let loc = s.buildable.loc in
   let source_rule =
     let theories =
       let open Resolve.Memo.O in
-      let* theory = theory in
-      let+ theories = cctx.theories_deps in
+      let+ theory = theory
+      and+ theories = theories_deps in
       theory :: theories
     in
+    let open Action_builder.O in
+    let* theories = Resolve.Memo.read theories in
     source_rule ~sctx theories
   in
-  Memo.parallel_iter coq_modules
-    ~f:(setup_rule ~sctx ~loc cctx ~source_rule ~dir ~file_targets:[])
+  let coqc_dir = (Super_context.context sctx).build_dir in
 
-let setup_coqdoc_rules ~sctx ~dir ~cctx (s : Theory.t) coq_modules =
-  let loc, name = (s.buildable.loc, snd s.name) in
-  let rule =
-    let file_flags = Context.coqdoc_file_flags cctx in
-    fun mode ->
-      let* () =
-        coqdoc_rule cctx ~sctx ~mode ~theories_deps:cctx.theories_deps ~name
-          ~file_flags coq_modules
-        |> Super_context.add_rule ~loc ~dir sctx
-      in
-      Coqdoc_mode.directory mode cctx.dir name
-      |> Path.build |> Action_builder.path
-      |> Rules.Produce.Alias.add_deps (Coqdoc_mode.alias mode ~dir) ~loc
+  let* mode = select_native_mode ~sctx ~dir s.buildable in
+  Memo.parallel_iter coq_modules
+    ~f:
+      (setup_coqdep_rule ~sctx ~dir ~loc ~theories_deps ~wrapper_name
+         ~use_stdlib ~source_rule ~ml_flags ~mlpack_rule)
+  >>> Memo.parallel_iter coq_modules
+        ~f:
+          (setup_coqc_rule ~loc ~dir ~sctx ~file_targets:[] ~stanza_flags
+             ~coqc_dir ~theories_deps ~mode ~wrapper_name ~use_stdlib ~ml_flags
+             ~theory_dirs)
+  >>> setup_coqdoc_rules ~sctx ~dir ~theories_deps s coq_modules
```

Comment:
Mmm, why do we separate coqdep from coqdep rules, that seems to add an unnecessary iteration?

Reply:
Coqdoc rules are done for an entire theory rather than incrementally. It was like that before this PR.